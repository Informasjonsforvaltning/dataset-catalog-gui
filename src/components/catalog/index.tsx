import React, { memo, FC } from 'react';
import { compose } from 'redux';
import memoize from 'lodash/memoize';
import { resolve } from 'react-resolver';

import { getDatasetsCount } from '../../services/api/registration-api/datasets';

import CatalogItem from '../catalog-item';

interface ExternalProps {
  catalogId: string;
  type: string;
  isReadOnly?: boolean;
  disabled: boolean;
}

interface Props extends ExternalProps {
  itemsCount?: number;
}

const Catalog: FC<Props> = ({
  catalogId,
  type,
  itemsCount,
  isReadOnly,
  disabled
}) => {
  const getLinkUri = () => `/catalogs/${catalogId}/${type}`;

  const linkUri = getLinkUri();

  return (
    <CatalogItem
      linkUri={linkUri}
      key={catalogId}
      type={type}
      itemsCount={itemsCount}
      isReadOnly={isReadOnly}
      disabled={disabled}
    />
  );
};

const memoizedGetDatasetsCount = memoize(getDatasetsCount);

const mapProps = {
  itemsCount: ({ type, catalogId, itemsCount }: any) => {
    switch (type) {
      case 'datasets': {
        return memoizedGetDatasetsCount(catalogId);
      }
      default:
        return itemsCount;
    }
  }
};

export default compose<FC<ExternalProps>>(memo, resolve(mapProps))(Catalog);
