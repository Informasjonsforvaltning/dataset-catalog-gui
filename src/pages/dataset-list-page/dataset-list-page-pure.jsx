import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import localization from '../../services/localization';
import { FormCatalog } from './form-catalog/form-catalog';
import { DatasetItemsList } from './items-list/dataset-item-list.component';
import { AlertMessage } from '../../components/alert-message/alert-message.component';
import './dataset-list-page.scss';

export const DatasetsListPagePure = props => {
  const {
    isReadOnly,
    catalogId,
    catalog,
    datasetItems,
    location,
    dispatchEnsureData,
    onClickCreateDataset
  } = props;

  useEffect(() => dispatchEnsureData(catalogId), [catalogId]);

  return (
    <div className="container" id="content">
      {_.get(location, ['state', 'confirmDelete'], false) && (
        <div className="row mb-5">
          <div className="col-12">
            <AlertMessage type="success">
              {localization.formStatus.type.dataset}{' '}
              {localization.formStatus.confirmDeleted}
            </AlertMessage>
          </div>
        </div>
      )}
      <div className="row">
        <div className="col-12" />
      </div>
      <div className="row mb-2 mb-md-5">
        {_.get(catalog, ['items', catalogId]) && (
          <div className="col-12 fdk-reg-datasets-list">
            <FormCatalog catalogId={catalogId} />
            <DatasetItemsList
              isReadOnly={isReadOnly}
              catalogId={catalogId}
              datasetItems={datasetItems}
              defaultEmptyListText={localization.listItems.missingDatasetItems}
              onClickCreateDataset={onClickCreateDataset}
            />
          </div>
        )}
      </div>
    </div>
  );
};

DatasetsListPagePure.defaultProps = {
  catalog: null,
  datasetItems: null,
  location: null,
  dispatchEnsureData: _.noop,
  onClickCreateDataset: _.noop
};

DatasetsListPagePure.propTypes = {
  catalogId: PropTypes.string.isRequired,
  catalog: PropTypes.object,
  datasetItems: PropTypes.array,
  location: PropTypes.object,
  dispatchEnsureData: PropTypes.func,
  onClickCreateDataset: PropTypes.func
};
