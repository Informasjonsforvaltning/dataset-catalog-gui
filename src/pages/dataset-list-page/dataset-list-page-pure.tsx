import React, { FC, memo, useEffect, useState } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import _ from 'lodash';
import AddIcon from '@material-ui/icons/Add';
import SearchIcon from '@material-ui/icons/Search';

import localization from '../../services/localization';
import { FormCatalog } from './form-catalog/form-catalog';
import DatasetItemsList from './items-list/dataset-item-list.component';
import { AlertMessage } from '../../components/alert-message/alert-message.component';
import SC from './styled';
import withDatasets, {
  Props as SearchProps
} from '../../components/with-datasets';

import { SearchRequest, Dataset } from '../../types';
import { SearchType } from '../../types/enums';

interface Props extends RouteComponentProps, SearchProps {
  catalogId: string;
  isReadOnly: boolean;
  catalog: any;
  datasetItems: Dataset[];
  dispatchEnsureData: (catalogId: string) => void;
  onClickCreateDataset: (catalogId: string) => void;
}

export const DatasetsListPagePure: FC<Props> = ({
  isReadOnly,
  catalogId,
  catalog,
  datasetItems,
  dispatchEnsureData,
  onClickCreateDataset,
  location,
  searchActions: { datasetsRequested: requestSearch }
}) => {
  useEffect(() => dispatchEnsureData(catalogId), [catalogId]);

  const [searchQuery, setSearchQuery] = useState('');

  useEffect(() => {
    const searchRequest: SearchRequest = {
      searchType: SearchType.DATASET,
      catalogIDs: [catalogId],
      query: searchQuery
    };

    requestSearch(searchRequest);
  }, [searchQuery]);

  return (
    <SC.DatasetListPage>
      {_.get(location, ['state', 'confirmDelete'], false) && (
        <AlertMessage type="success">
          {localization.formStatus.type.dataset}{' '}
          {localization.formStatus.confirmDeleted}
        </AlertMessage>
      )}
      {catalog.items?.[catalogId] && (
        <FormCatalog catalogId={catalogId} isReadOnly={isReadOnly} />
      )}
      <SC.ListActions>
        {!isReadOnly && (
          <SC.CreateButton
            type="button"
            onClick={() => onClickCreateDataset(catalogId)}
          >
            <AddIcon />
            {localization.datasets.list.btnNewDataset}
          </SC.CreateButton>
        )}
        <SC.SearchBox role="search">
          <SC.SearchField
            type="text"
            placeholder="SÃ¸k etter datasettbeskrivelse"
            value={searchQuery}
            onChange={({ target: { value } }) => setSearchQuery(value)}
          />
          <SearchIcon />
        </SC.SearchBox>
      </SC.ListActions>
      <DatasetItemsList
        catalogId={catalogId}
        searchSort={searchQuery !== ''}
        datasetItems={datasetItems}
      />
    </SC.DatasetListPage>
  );
};

export default withDatasets(withRouter(memo(DatasetsListPagePure)));
