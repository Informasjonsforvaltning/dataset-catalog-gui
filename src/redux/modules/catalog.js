import _ from 'lodash';
import { registrationApiGet } from '../../services/api/registration-api/host';
import { reduxFsaThunk } from '../../lib/redux-fsa-thunk';
import { catalogPath } from '../../services/api/registration-api/catalogs';

const CATALOG_REQUEST = 'CATALOG_REQUEST';
export const CATALOG_SUCCESS = 'CATALOG_SUCCESS';
const CATALOG_FAILURE = 'CATALOG_FAILURE';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export const fetchCatalogIfNeeded = catalogId => (dispatch, getState) =>
  shouldFetch(_.get(getState(), ['catalog', 'meta', catalogId])) &&
  dispatch(
    reduxFsaThunk(() => registrationApiGet(catalogPath(catalogId)), {
      onBeforeStart: { type: CATALOG_REQUEST, meta: { catalogId } },
      onSuccess: { type: CATALOG_SUCCESS, meta: { catalogId } },
      onError: { type: CATALOG_FAILURE, meta: { catalogId } }
    })
  );

const initialState = {};

export default function catalog(state = initialState, action) {
  switch (action.type) {
    case CATALOG_REQUEST:
      return {
        ...state,
        meta: {
          [action.meta.catalogId]: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case CATALOG_SUCCESS:
      return {
        ...state,
        items: {
          [action.meta.catalogId]: action.payload
        },
        meta: {
          [action.meta.catalogId]: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case CATALOG_FAILURE:
      return {
        ...state,
        meta: {
          [action.meta.catalogId]: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}
