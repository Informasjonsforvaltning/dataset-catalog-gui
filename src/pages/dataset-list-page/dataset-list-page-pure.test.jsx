import React from 'react';
import { shallow } from 'enzyme';
import { DatasetsListPagePure } from './dataset-list-page-pure';
import catalog from './__fixtures/catalog';

let defaultProps;
let wrapper;

beforeEach(() => {
  const fetchCatalogIfNeeded = jest.fn();
  const fetchDatasetsIfNeeded = jest.fn();
  defaultProps = {
    catalogId: catalog.item.id,
    catalog,
    fetchCatalogIfNeeded,
    fetchDatasetsIfNeeded
  };
  wrapper = shallow(<DatasetsListPagePure {...defaultProps} />);
});

test('should render DatasetsListPagePure correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
