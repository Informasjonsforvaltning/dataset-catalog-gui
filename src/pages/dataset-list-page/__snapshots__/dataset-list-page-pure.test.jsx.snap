// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetsListPagePure correctly 1`] = `
<div
  className="container"
  id="content"
>
  <div
    className="row"
  >
    <div
      className="col-12"
    />
  </div>
  <div
    className="row mb-2 mb-md-5"
  />
</div>
`;
